include/beast/http/detail/basic_parser.hpp

```diff
@@ -0,0 +1,388 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_HTTP_DETAIL_BASIC_PARSER_HPP
+#define BEAST_HTTP_DETAIL_BASIC_PARSER_HPP
+
+#include <boost/system/error_code.hpp>
+#include <boost/utility/string_ref.hpp>
+#include <array>
+#include <cstdint>
+
+namespace beast {
+namespace http {
+namespace detail {
+
+// '0'...'9'
+inline
+bool
+is_digit(char c)
+{
+    return c >= '0' && c <= '9';
+}
+
+inline
+bool
+is_token(char c)
+{
+    /*  token = 1*<any CHAR except CTLs or separators>
+        CHAR  = <any US-ASCII character (octets 0 - 127)>
+        sep   = "(" | ")" | "<" | ">" | "@"
+              | "," | ";" | ":" | "\" | <">
+              | "/" | "[" | "]" | "?" | "="
+              | "{" | "}" | SP | HT
+    */
+    static std::array<char, 256> constexpr tab = {{
+        0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0, // 0
+        0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0, // 16
+        0, 1, 0, 1,  1, 1, 1, 1,  0, 0, 1, 1,  0, 1, 1, 0, // 32
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 0, 0,  0, 0, 0, 0, // 48
+        0, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 64
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 0,  0, 0, 1, 1, // 80
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 96
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 0,  1, 0, 1, 0, // 112
+    }};
+    return tab[static_cast<std::uint8_t>(c)] != 0;
+}
+
+inline
+bool
+is_text(char c)
+{
+    // TEXT = <any OCTET except CTLs, but including LWS>
+    static std::array<char, 256> constexpr tab = {{
+        0, 0, 0, 0,  0, 0, 0, 0,  0, 1, 0, 0,  0, 0, 0, 0, // 0
+        0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0, // 16
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 32
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 48
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 64
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 80
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 96
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 0, // 112
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 128
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 144
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 160
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 176
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 192
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 208
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1, // 224
+        1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1,  1, 1, 1, 1  // 240
+    }};
+    return tab[static_cast<std::uint8_t>(c)] != 0;
```

Comment:
A bitset may be cache friendlier.

     static std::bitset<256> tab {
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "01111111111111111111111111111111"
        "11111111111111111111111111111111"
        "11111111111111111111111111111111"
        "00000000000000000000001000000000"
    };